tximeta for "other" species Part 2: Add 00 Rmd notebook (#379)
AlexsLemonade/training-modules: RNA-seq/00c-tximeta_other_species.Rmd

---
title: "Gene-level summarization with `tximeta` for non-mouse or non-human"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

**CCDL 2021**

In this notebook, we'll import the transcript expression quantification output from `salmon quant` using the [`tximeta`](https://bioconductor.org/packages/release/bioc/html/tximeta.html) package but for a species not human or mouse. 

The steps for human and mouse are more straightforward and are covered in `02-gastric_cancer_tximeta-live.Rmd`. 
The main difference for using `tximeta` for other species is that we need to set up the reference data in the form of a `linkedTxome` before running `tximeta()` function. 

![](diagrams/rna-seq_5.png)

For more information about `tximeta`, see [this excellent vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/tximeta/inst/doc/tximeta.html) from Love _et al_.

> ## Libraries and functions

Author: Not sure if there's a recommended way to let people know that much of this is the same as usual.

# Set up file path to the salmon index for the species we are using
index_dir <- file.path("index", "Danio_rerio", "short_index")
```

These directories and files are for pointing to the `quant.sf` files. 

```{r}
# directory where the data are located
data_dir <- file.path("data", "zebrafish-cortisol")

# directory where the quant files are located, each sample is its own
# directory
quant_dir <- file.path(data_dir, "salmon_quant")

# create a directory to hold the tximport results if it doesn't exist yet
txi_dir <- file.path(data_dir, "tximport")
if (!dir.exists(txi_dir)) {
  dir.create(txi_dir, recursive = TRUE)
}
```

```{r input-names}
# the quant files themselves
sf_files <- list.files(quant_dir, recursive = TRUE, full.names = TRUE,
                       pattern = "quant.sf")
```

```{r metadata-file}
# sample metadata file
meta_file <- file.path(data_dir, "zebrafish-cortisol_metadata.tsv")
```

**Output**

```{r output-names}
# Name the output zebrafish-cortisol_tximport.RDS and use the directory created 
# above as the rest of the path
txi_out_file <- file.path(txi_dir, "zebrafish-cortisol_tximeta.RDS")
```

## Set up reference data for a non-human, non-mouse species 

> Here we will set this up for zebrafish, _Danio rerio_.

Reviewer: We should add some context here about the ensembl versions. Why are we using release 98 here? Because that was the version used when we made the salmon index. If they make their own index, they should know that this would be different.

We might actually want to provide a bit of info here about building an index; while we provide some, we might not have what somebody wants. For the ones that we do provide, we might think about having a table with links to the relevant metadata, rather than making people search for it.

A question I have: Is it possible to find the version info from the index itself, and if so, how? If it is possible, this is information we might want to provide.

Author: There does not appear to be an easy way to extract the original genome info from the salmon index. So I just tried to descibe this issue and link to our github `training-txome-prep` repo where relevant.

You can look at Ensembl's ftp server to see the full list of species and fasta file options: http://ftp.ensembl.org/pub/release-98/fasta/

If you are using a species other than zebrafish, click through the file paths to find the file you want to use for annotation. 

> We will store the `base_url` where all the species FASTA files are located and then the second part of the url is species-specific which you may have to replace. 

Reviewer: What if they are not getting anotation data from ensembl? We don't have to cover that use case, but we should be clear about what these instructions do and don't cover.

Author: Hopefully my updated explanations elude to this enough without getting caught on a tangent.

```{r}
fasta_base_url <- "ftp://ftp.ensembl.org/pub/release-98/fasta/"

> fasta_files <- c(

Reviewer: I suggest `urls` here to distinguish the fact that these are not files on this machine.

Is it possible to use downloaded files? If so, what is the tradeoff between links and local files?

```suggestion
fasta_urls <- c(
```

  paste0(
    fasta_base_url, 
    # Replace the file path below to the species specific fasta cDNA file
    "danio_rerio/cdna/Danio_rerio.GRCz11.cdna.all.fa.gz" 
    ),
  paste0(
    fasta_base_url, 
    # Replace the non-coding RNA fasta file path so that non-coding genes are also annotated

>     "danio_rerio/ncrna/Danio_rerio.GRCz11.ncrna.fa.gz" 

Reviewer: Are these in our index? I don't think so, but I don't know for sure.

Author: They are not. It does not appear including them hurt anything but probably this information in this file doesn't actually go anywhere since its not included in our index in the first place.

Author: I dropped the ncRNA file because I don't think it was doing anything besides adding an extra step.

coldata <- data.frame(files = sf_files, 
                      names = sample_names)
```

We have more information about these samples stored in the metadata file that we will also want stored in `coldata`. 
Let's read in the sample metadata from the TSV file and do a bit of set up. 
We really only want two columns from this metadata file, so we will use a `dplyr::select()` to keep only those. 

```{r sample_meta_df}
# Read in the sample metadata TSV file and have a look
sample_meta_df <- readr::read_tsv(meta_file) %>% 
  # Select the two columns we need here
  dplyr::select(refinebio_accession_code, refinebio_title)

# Take a look at the preview
sample_meta_df
```

We'll want this information to be added to the `coldata`, which we can do by using a join function to match up the rows between the two data frames and combine them. 

```{r join-sample_meta_df}
coldata <- coldata %>%
  dplyr::inner_join(sample_meta_df, by = c("names" = "refinebio_accession_code"))

coldata
```

## Import expression data with `tximeta`

Using the `coldata` data frame and the `linkedTxome` we set up, now run the `tximeta()` per usual.

```{r tximeta}
txi_data <- tximeta(coldata)
```

> If you haven't created this linkedTxome before, in the console window, you'll be asked: 

Reviewer: Is there any way to trigger this step before calling `tximeta()`? I would be confused having thought that I had already created the `linkedTxome` when I called `makeLinkedTxome()`.

If the answer is to the above question is `no` I would suggest the following edit:
```suggestion
If you haven't used this linkedTxome before, in the console window, you'll be asked: 
```

coldata <- coldata %>%
  dplyr::inner_join(sample_meta_df, by = c("names" = "refinebio_accession_code"))

coldata
```

## Set up reference data for a non-human, non-mouse species 

```{r}
indexDir <- file.path(dir, "Dm.BDGP6.22.98_salmon-0.14.1")
```

You can go to Ensembl to see the full list of species: http://ftp.ensembl.org/pub/release-98/fasta/
Click through the file paths to find the file you want to use for annotation. 
Right click on the file and copy its link address
We will store the `base_url` where all the species FASTA files are located and then the second part of the url is species specific. 

```{r}
base_url <- "ftp://ftp.ensembl.org/pub/release-98/fasta/"

fasta_ftp <- c(
  paste0(base_url, "danio_rerio/cdna/Danio_rerio.GRCz11.cdna.all.fa.gz"),
  paste0(base_url, "danio_rerio/ncrna/Danio_rerio.GRCz11.ncrna.fa.gz")
)
```

Now you can follow a similar series of steps to obtain the Gene Transfer File (GTF) using this address: 
http://ftp.ensembl.org/pub/release-103/gtf/

```{r}
base_url <- "ftp://ftp.ensembl.org/pub/release-103/gtf/"

gtf_path <- paste0(base_url, "danio_rerio/Danio_rerio.GRCz11.103.gtf.gz")
```

Create a linkedTxome with the 

```{r}

> makeLinkedTxome(indexDir = index_dir,

Author: I am unsure if these link-data.sh is working right. I'm getting this error when trying to use the zebrafish salmon index: 
```
Error: lexical error: invalid char in json text. index/Danio_rerio/header.json (right here) ------^
```

Note that the FASTA files used here should be the same as what you used to build your salmon index. 
If you used multiple FASTA files to build your index, you can make this into a vector and include multiple FASTA files here as well. 

```{r}
fasta_urls <- paste0(
  fasta_base_url, 
  # Replace the file path below to the species-specific fasta cDNA file
  "danio_rerio/cdna/Danio_rerio.GRCz11.cdna.all.fa.gz" 
  )
```

### GTF file

Now we can follow a similar series of steps to set up the GTF file's URL for our species using this address: 
http://ftp.ensembl.org/pub/release-95/gtf/

Again, if you are using a different release, the `release-95` part below will need to be altered. 

```{r}
# This base URL will not need to change for other species of the same release
gtf_base_url <- "ftp://ftp.ensembl.org/pub/release-95/gtf/"
```

Similarly, the second part of this URL will need to be replaced with your species' equivalent file path (if you are not using zebrafish). 

```{r}
gtf_url <- paste0(
  gtf_base_url, 
  # Replace the file path below to the species-specific gtf file

>   "danio_rerio/Danio_rerio.GRCz11.103.gtf.gz"

Author: Why did this still work? (Or did it not and tximeta still ran anyway)?

>   "danio_rerio/Danio_rerio.GRCz11.103.gtf.gz"

Reviewer: Not sure, because I didn't test it wrong, but my guess (from some of the status messages) is that it doesn't actually download the gtf if it can find the ensembl annotation through `AnnotationDbi`. I tried to see what would happen if  I left off fasta and/or gtf files, and that failed at the `makeLinkedTxome` step, but for the full case, when you actually get to `tixmeta` it says something about downloading from AnnotationDbi, so I think it shortcuts when it can.

>   "danio_rerio/Danio_rerio.GRCz11.103.gtf.gz"

Author: SOooo even after the url is changed to release-95, `tximeta()` pulls from a LinkedTxome that still has mixed releases: 
<img width="866" alt="Screen Shot 2021-02-25 at 8 14 42 AM" src="https://user-images.githubusercontent.com/23458084/109158670-ad5c0280-7741-11eb-9163-b12907e9be70.png">

Next, I tried clearing all the LinkedTxomes and re-creating one that has the files with all the releases matching. But in a weirder twist, it now does not recognize that the created matching-release linkedTxome for the current zebrafish dataset. So I'm not sure how it determines what linkedTxome works for what dataset. 

Going to dig a tad more because this is weird.

>   "danio_rerio/Danio_rerio.GRCz11.103.gtf.gz"

Author: Figured it out with jashapiro's help. The salmon_index in my own directory was the wrong species leftover from when I was editing `setup/link-data.sh` and had an error in it. Though I had re-run `link-data.sh` that wasn't enough to fix it because it needed to be deleted. So basically I was sending the `linkedTxome` to a `Homo_sapiens` index that I thought was `Danio_rerio`. Once I fixed that and re-made the `linkedTxome` now `tximeta` runs properly. 

Bigger picture lessons learned from this about tximeta: 1) the linkedTxome won't complain if the species are different in the files you provide it. 2) tximeta might run with a linkedTxome that has different releases if you have one that matches.

}
```

```{r input-names}
# Identify the quant files themselves and store them in a vector, `sf_files`
sf_files <- list.files(quant_dir, recursive = TRUE, full.names = TRUE,
                       pattern = "quant.sf")
```

```{r metadata-file}
# sample metadata file
meta_file <- file.path(data_dir, "zebrafish-cortisol_metadata.tsv")
```

**Output**

```{r output-names}
# Name the output zebrafish-cortisol_tximport.RDS and use the directory created 
# above as the rest of the path
txi_out_file <- file.path(txi_dir, "zebrafish-cortisol_tximeta.RDS")
```

## File names

```{r sf_files}
# Let's look at the full path for the quant.sf files
sf_files
```

Let's extract the _sample_ names from the **file paths** using the `stringr` package.

```{r sample_names}
sample_names <- stringr::word(sf_files, -2, sep = "/")
sample_names
```

## Set up metadata

```{r names_sf_files}
coldata <- data.frame(files = sf_files, 
                      names = sample_names)
```

We have more information about these samples stored in the metadata file that we will also want stored in `coldata`. 
Let's read in the sample metadata from the TSV file and do a bit of set up. 

> We really only want two columns from this metadata file, so we will use a `dplyr::select()` to keep only those. 

Author: I was trying to make sure the tximeta data output we made was still reasonably useable in case participants wanted to do some further analyses, but I didn't bother doing the splitting. 

It would be just one line. And now I'm thinking if a participant does use this data they are going to ask us how to do this anyway, so might as well do it now.

txi_data <- tximeta(coldata)
```

If you haven't used this `linkedTxome` before, you may be asked (in the console window) about whether to create a directory. You can feel free to say yes.

> If we did not have a `linkedTxome` for our dataset we tried to run `tximeta()` we would get the following message:

Reviewer: ```suggestion
If we had not created a `linkedTxome` for the salmon index we used to quantify our data set, `tximeta()` we would give the following message:
```
